internal/trie/node/decode.go

```diff
@@ -108,41 +102,37 @@ func decodeBranch(reader io.Reader, header byte) (branch *Branch, err error) {
 
 		// Handle inlined leaf nodes.
 		const hashLength = 32
-		if Type(hash[0]>>6) == LeafType && len(hash) < hashLength {
+		nodeTypeHeaderByte := hash[0] >> 6
+		if nodeTypeHeaderByte == leafHeader && len(hash) < hashLength {
 			leaf, err := decodeLeaf(bytes.NewReader(hash[1:]), hash[0])
 			if err != nil {
 				return nil, fmt.Errorf("%w: at index %d: %s",
 					ErrDecodeValue, i, err)
 			}
-			branch.AddDescendants(1)
-			branch.Children[i] = leaf
+			node.Descendants++
+			node.Children[i] = leaf
 			continue
 		}
 
-		branch.AddDescendants(1)
-		branch.Children[i] = &Leaf{
+		node.Descendants++
+		node.Children[i] = &Node{
 			HashDigest: hash,
 		}
 	}
 
-	branch.Dirty = true
+	node.Dirty = true
 
-	return branch, nil
+	return node, nil
 }
 
 // decodeLeaf reads and decodes from a reader with the encoding specified in lib/trie/node/encode_doc.go.
-func decodeLeaf(reader io.Reader, header byte) (leaf *Leaf, err error) {
-	nodeType := Type(header >> 6)
-	if nodeType != LeafType {
-		return nil, fmt.Errorf("%w: %d", ErrNodeTypeIsNotALeaf, nodeType)
-	}
-
-	leaf = &Leaf{
```

Comment:
same comment as above in decodeBranch

Reply:
This function is not exported and is used in only one place, I would still like to remove that check.
I recall I added it previously to detect decoding errors I had whilst testing it, I probably forgot to remove it.

On top of that:

- Tests (unit, e2e, fuzzing) should detect such bad use of this function
- The header (variant + partial key length) decoding is pushed before `decodeBranch` / `decodeLeaf`. This is even more robust in the next PR #2530